### PR TITLE
update solochain to use #[runtime]

### PR DESCRIPTION
**Pull request by Jan-Jan**
# Description

* This is part of [issue 5242](https://github.com/paritytech/polkadot-sdk/issues/5242), specifically getting solochain to use `#[frame::runtime]`
* Furthermore, reinforced the convention of `Template` instead of `TemplateModule`

## Integration

* Should be integrated into the `solochain` template and documentation

## Review Notes

* Refactored `solochain` template from [construct_runtime!](https://paritytech.github.io/polkadot-sdk/master/frame_support/macro.construct_runtime.html) to [#[runtime]](https://paritytech.github.io/polkadot-sdk/master/frame_support/attr.runtime.html).
* AFAIU `Template` is our new convention, and preferred over `TemplateModule`.

# Out of scope

* The [#[runtime]](https://paritytech.github.io/polkadot-sdk/master/frame_support/attr.runtime.html) documentation is still very rudimentary, and should ideally be expanded to explain the macro, both what it does and the input options.
* Furthermore, suggest update  [#[runtime]](https://paritytech.github.io/polkadot-sdk/master/frame_support/attr.runtime.html) documentation to replace `#[crate::runtime]` with `#[frame_support::runtime]`

### CODE REVIEW

**Review by bkchr — APPROVED** [review]

**Review by gupnik — APPROVED** [review]

File: templates/solochain/pallets/template/src/mock.rs [hunk 1]
```diff
@@ -4,14 +4,33 @@ use sp_runtime::BuildStorage;
...
+		RuntimeEvent,
+		RuntimeError,
+		RuntimeOrigin,
+		RuntimeFreezeReason,
+		RuntimeHoldReason,
+		RuntimeSlashReason,
+		RuntimeLockId,
+		RuntimeTask
+	)]
+	// pub struct Runtime;
```
Comment (ggwpez) [inline templates/solochain/pallets/template/src/mock.rs:23]:
```suggestion
```

File: templates/solochain/pallets/template/src/mock.rs [hunk 2]
```diff
@@ -4,14 +4,33 @@ use sp_runtime::BuildStorage;
...
+	pub struct Test;
+
+	// Use the concrete pallet type
```
Comment (ggwpez) [inline templates/solochain/pallets/template/src/mock.rs:26]:
```suggestion
```
I think its obvious whats happening.

File: templates/solochain/pallets/template/src/mock.rs [hunk 3]
```diff
@@ -4,14 +4,33 @@ use sp_runtime::BuildStorage;
...
+	#[runtime::pallet_index(0)]
+	pub type System = frame_system::Pallet<Test>;
+
+	// // Use path to the pallet
```
Comment (ggwpez) [inline templates/solochain/pallets/template/src/mock.rs:30]:
```suggestion
```

**Review by ggwpez — APPROVED** [review]

**Review by shawntabrizi — APPROVED** [review]

File: templates/solochain/pallets/template/src/mock.rs [hunk 4]
```diff
@@ -4,14 +4,30 @@ use sp_runtime::BuildStorage;
...
+		RuntimeCall,
+		RuntimeEvent,
+		RuntimeError,
+		RuntimeOrigin,
+		RuntimeFreezeReason,
+		RuntimeHoldReason,
+		RuntimeSlashReason,
+		RuntimeLockId,
+		RuntimeTask
+	)]
```
Comment (shawntabrizi) [inline templates/solochain/pallets/template/src/mock.rs:22]: not sure the philosophy we want to follow, but the template does not need all these derives. i could see an argument to use it either way

**Comment by shawntabrizi** [comment]
bot fmt

File: templates/solochain/pallets/template/src/mock.rs (hunk 4 above)
Comment (Jan-Jan) [reply templates/solochain/pallets/template/src/mock.rs:22]: I'm pro reducing this, but I cannot find the documentation explaining what each does. My personal preference would be to exclude them, and explain in the template why we are excluding them.

Comment (shawntabrizi) [reply templates/solochain/pallets/template/src/mock.rs:22]: if the type isnt being used anywhere in the file, you can pretty safely delete it.

Comment (kianenigma) [reply templates/solochain/pallets/template/src/mock.rs:22]: @Jan-Jan seems like the PR got merged without this being addressed. Looks like a good follow up. You can also look at the `mock.rs` in the parachain template and fix this there too.

**Pull request merged**